src/timer/timer.3
.\"(c) Copyright 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: timer.3,v 1.2 2003/06/17 05:10:55 seth Exp $
.TH TIMER 3X "20 April 1993"
.SH NAME
timer_create, timer_destroy, timer_start, timer_stop, timer_block, timer_unblock, timer_expirations, timer_block_type, timer_unblock_type - timer management functions
.SH SYNOPSIS
.LP
.nf
.ft B
#include "timer.h"
.LP
.ft B
enum timer_types { TIMER_REAL, TIMER_VIRTUAL, TIMER_PROF } ;
enum timer_timetypes { TIMER_ABSOLUTE, TIMER_RELATIVE } ;
.LP
.ft B
extern int timer_errno ;
.LP
.ft B
timer_h timer_create( type, flags, errnop )
enum timer_types type ;
int flags ;
int *errnop ;
.LP
.ft B
void timer_destroy( handle )
timer_h handle ;
.LP
.ft B
int timer_start( handle, itvp, time_type, action )
timer_h handle ;
struct itimerval *itvp ;
enum timer_timetypes time_type ;
struct timer_action *action ;
.LP
.ft B
void timer_stop( handle )
timer_h handle ;
.LP
.ft B
void timer_block( handle )
timer_h handle ;
.LP
.ft B
void timer_unblock( handle )
timer_h handle ;
.LP
.ft B
unsigned timer_expirations( handle )
timer_h handle ;
.LP
.ft B
void timer_block_type( type )
enum timer_types type ;
.LP
.ft B
void timer_unblock_type( type )
enum timer_types type ;
.SH DESCRIPTION
.LP
This library provides support for multiple timers of various types by
multiplexing the timers provided by the operating system.
An action is associated with each timer.
The action can either be a function that will be invoked when the
timer expires,
or it can be an integer variable that will either be set to
.SM TRUE
when the timer expires or it will be increased for every expiration
of the timer (it is up to the user to clear the variable after inspecting
its value).
If a function is used, then it is possible that while that function is running,
more timers may expire and their functions be called
(i.e. timer interrupts are not blocked while the function is running).
The timer-action association can change through the lifetime of the timer.
The timer types supported by this library depend on the operating system.
.\" *********************** timer_create *****************************
.LP
.B timer_create()
creates a timer and returns a handle to it. When the timer is created, it
is inactive. Possible values for \fItype\fP are:
.TP 18
.SB TIMER_REAL
timers of this type are wall-clock timers
.TP
.SB TIMER_VIRTUAL
timers of this type are running only when the process is running
.TP
.SB TIMER_PROF
timers of this type are running only when the process is running
or the system is doing work on behalf of the process.
.LP
Possible values for \fIflags\fP are formed by ORing any of the following
constants:
.TP 20
.SB TIMER_RETURN_ERROR
if an error
occurs while handling \fIthis\fP timer an error indication will be returned
(the default is program termination)
.LP
You can use the constant
.B TIMER_NOFLAGS
to specify no flags at all.
The argument
.I errnop
is a pointer to a variable where error values will be placed
(the default, in case
.I errnop
is
.SM NULL,
is \fItimer_errno\fP).
.\" *********************** timer_destroy *****************************
.LP
.B timer_destroy()
destroys the timer specified by \fIhandle\fP.
If the timer is active, it is deactivated first.
.\" *********************** timer_start *****************************
.LP
.B timer_start()
activates the timer specified by
.I handle.
The expiration time is determined by
.I itvp
and
.I time_type.
.I time_type
can be either
.B TIMER_ABSOLUTE
or
.B TIMER_RELATIVE.
Timers of the former
type expire at the time
specified by
.I "itvp->it_value"
(if that time is before the
current time then the timer expires immediately), while the expiration
time of timers of the latter type expire at
.I "current time + itvp->it_value."
.I "itvp->it_interval"
is the interval between subsequent expirations of the timer after the
original expiration.
The
.I action
argument determines the action to be taken when the timer expires.
.LP
.I "struct timer_action"
is defined as follows:
.sp 1
.PD .1v
.RS
.nf
struct timer_action
{
.RS
.IP "int" 15
ta_flags ;
.IP "void"
(*ta_func)() ;
.IP void
*ta_arg ;
.IP jmp_buf
ta_env ;
.RE
} ;
.RE
.PD
.fi
.LP
The type of action taken when a timer expires depends on
.I ta_func
and
.I ta_arg.
.RS
.IP "Case 1: ta_func != NULL"
The function will be invoked with 2 arguments: the handle of the expired
timer and
.I ta_arg.
.I "Such functions should not use longjmp."
.IP "Case 2: ta_func == NULL && ta_arg != NULL"
.I ta_arg
is assumed to be a pointer to an integer variable. This variable will
either be set to
.SM TRUE
or its value will be increased.
.IP "Case 3: ta_func == NULL && ta_arg == NULL"
nothing happens.
.RE
.sp 1
.I ta_flags
is formed by ORing any of the following constants:
.TP 20
.SB TIMER_INC_VAR
When the action associated with the timer is an integer variable then
the variable will be increased by the number of timer expirations (the
default is for the variable to be set to 1).
.TP
.SB TIMER_BLOCK_ALL
When the action associated with the timer is a function, all timer
interrupts will be blocked while the function is running.
.TP
.SB TIMER_BLOCK_SAME
When the action associated with the timer is a function, all timers
of the same type will be blocked while the function is running.
.TP
.SB TIMER_LONGJMP
A longjmp(3) will be performed using
.I ta_env
after the action associated with the timer is performed.
.LP
You can use the constant
.B TIMER_NOFLAGS
to specify no flags at all.
.\" *********************** timer_stop *****************************
.LP
.B timer_stop()
stops the specified timer. The timer becomes inactive and it cannot
be restarted for the time that was remaining until expiration.
.\" *********************** timer_block *****************************
.LP
.B timer_block()
blocks the specified timer. The timer is still active and may expire
but its action will not be executed until the
timer is unblocked.
.\" *********************** timer_unblock *****************************
.LP
.B timer_unblock()
unblocks the specified timer. If the timer is past its expiration time,
the associated action will be executed immediately.
.\" *********************** timer_expirations *****************************
.LP
.B timer_expirations()
should be used by the user-specified function to find out
the number of times the timer expired until the function was called.
Multiple expirations are possible for a timer that expires periodically.
.\" *********************** timer_block_type *****************************
.LP
.B timer_block_type()
blocks all timers of the specified \fItype\fP.
This also includes any timers started
.I after
this function is invoked.
.\" *********************** timer_unblock_type *****************************
.LP
.B timer_unblock_type()
unblocks all timers of the specified \fItype\fP.
.\" *********************** notes *****************************
.SH NOTES
.LP
Any of the timer operations can be used on any timer handle at any time.
Specifically, the operations can be used from within the functions
invoked when the timers expire.
.LP
The function associated with a timer is not
invoked for each expiration of that timer.
It is possible for a timer to expire
multiple times while its function is running (because timers can
be scheduled to expire periodically), or while other timer functions
are running. The timer function can find out how many times its timer
has expired by using the
.B timer_expirations()
operation.
.LP
It is guaranteed that the function associated with a specific timer will
not be called recursively if that timer expires multiple times.
.LP
The order of execution of timer-associated functions for
timers that expire at the same time is undefined.
.SH "RETURN VALUES"
.LP
Values for
.I timer_errno
will be stored in the user-specified variable if one was provided.
.LP
.B timer_create()
returns a timer handle if successful or
.SM NULL
if it fails.
.LP
.B timer_start()
returns
.B TIMER_OK
if successful or
.B TIMER_ERR
if it fails.
.LP
.B timer_expirations()
returns a positive (non-zero) number when invoked from a timer-associated
function; otherwise its return value is undefined.
.RE
.SH "ERRORS"
The following is a list of error codes which will be placed in
.I timer_errno
or the user-specified variable when a timer operation fails:
.RS
.TP 20
.SB TIMER_ENOMEM
A memory allocation request failed.
.TP
.SB TIMER_EBADTYPE
An unknown timer type was specified.
.TP
.SB TIMER_ESIGPROBLEM
A signal handler could not be installed.
.TP
.SB TIMER_EBADSTATE
The timer state does not allow this operation (for example, the timer is
running and the operation attempted was \fBtimer_start()\fP).
.TP
.SB TIMER_EBADTIME
The time value specified was negative.
.TP
.SB TIMER_ENOTAVAILABLE
The requested timer type is not available.
.TP
.SB TIMER_ECANTINSERT
The insertion of this timer in the queue of timers failed.
.TP
.SB TIMER_SIGPROBLEM
There was an error while trying to install a signal handler.
.SH "SEE ALSO"
setitimer(2), setjmp(3), longjmp(3)


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/timer/timer.c
/*
 * (c) Copyright 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */

static const char RCSid[] = "$Id: timer.c,v 1.3 2003/06/17 05:10:55 seth Exp $";
static const char version[] = VERSION;

#include <unistd.h>
#include <string.h>
#include "timemacros.h"
#include "impl.h"
#include "defs.h"
#include "clchack.h"

#define TIMER_H_NULL					((timer_h)NULL)


int timer_errno ;



/*
 * Create a timer of the specified type.
 * Returns a timer handle
 */
timer_h clc_timer_create(enum timer_types type, int flags, int *errnop)
{
	int		*errp = ( errnop != NULL ) ? errnop : &timer_errno ;
	timer_s	*tp ;

	if ( type != TIMER_REAL && type != TIMER_VIRTUAL && type != TIMER_PROF )
			HANDLE_ERROR( flags, TIMER_H_NULL, errp, TIMER_EBADTYPE,
							"TIMER timer_create: bad timer type\n" ) ;
		
	tp = TIMER_ALLOC() ;
	if ( tp == NULL )
	{
		*errp = TIMER_ENOMEM ;
		return( TIMER_H_NULL ) ;
	}

	tp->t_state = INACTIVE ;
	tp->t_act = IDLE ;
	tp->t_blocked = FALSE ;

	tp->t_errnop = errp ;
	tp->t_flags = flags & TIMER_CREATE_FLAGS ;
	tp->t_action.ta_func = NULL ;
	tp->t_action.ta_arg = NULL ;
	tp->t_action.ta_flags = TIMER_NOFLAGS ;
	if ( __ostimer_newtimer( tp, type ) == TIMER_ERR )
	{
		TIMER_FREE( tp ) ;
		return( TIMER_H_NULL ) ;
	}
	return( (timer_h) tp ) ;
}


void clc_timer_destroy(timer_h handle)
{
	timer_s	*tp = TP( handle ) ;

	__ostimer_blockall() ;

	if ( tp->t_state == TICKING )
	{
		__ostimer_remove( tp->t_ostimer, tp ) ;
		tp->t_state = DESTROYED ;
	}

	if ( tp->t_act == IDLE || tp->t_act == PENDING )
		TIMER_FREE( tp ) ;

	__ostimer_unblockall() ;
}


int clc_timer_start(timer_h handle, struct itimerval *itvp, enum timer_timetypes time_type, struct timer_action *ap)
{
	int					result ;
	int					ok_to_start ;
	timer_s				*tp	= TP( handle ) ;
	struct os_timer	*otp	= tp->t_ostimer ;

	__ostimer_blockall() ;

	/*
	 * We allow invoking timer_start from within the user-specified action
	 * after the timer has expired. However, we do not allow this for
	 * timers that have a t_interval (these timers stay at the TICKING state).
	 */
	ok_to_start = tp->t_state == INACTIVE &&
							( tp->t_act == IDLE || tp->t_act == INVOKED ) ;

	if ( ! ok_to_start )
	{
		__ostimer_unblockall() ;
		HANDLE_ERROR( tp->t_flags, TIMER_ERR, tp->t_errnop, TIMER_EBADSTATE,
			"TIMER timer_start: timer state does not allow this operation\n" ) ;
	}

	if ( itvp->it_value.tv_sec < 0 || itvp->it_value.tv_usec < 0 )
	{
		__ostimer_unblockall() ;
		HANDLE_ERROR( tp->t_flags, TIMER_ERR, tp->t_errnop, TIMER_EBADTIME,
			"TIMER timer_start: neg time value)\n" ) ;
	}

	tp->t_action = *ap ;
	tp->t_action.ta_flags &= TIMER_START_FLAGS ;

	result = __ostimer_add( otp, tp, itvp, time_type ) ;
	__ostimer_unblockall() ;
	return( result ) ;
}


void clc_timer_stop(timer_h handle)
{
	timer_s *tp = TP( handle ) ;

	__ostimer_blockall() ;

	if ( tp->t_state == TICKING )
	{
		__ostimer_remove( tp->t_ostimer, tp ) ;
		tp->t_state = INACTIVE ;
	}

	if ( tp->t_act == SCHEDULED )
		tp->t_act = INVOKED ;		/* to avoid the invocation */
	else if ( tp->t_act == PENDING )
		tp->t_act = IDLE ;

	tp->t_blocked = FALSE ;

	__ostimer_unblockall() ;
}


void clc_timer_block(timer_h handle)
{
	timer_s *tp = TP( handle ) ;

	__ostimer_blockall() ;

	if (tp->t_state == TICKING ||
	    (tp->t_state == INACTIVE &&
	     (tp->t_act == PENDING || tp->t_act == SCHEDULED)))
		tp->t_blocked = TRUE ;

	__ostimer_unblockall() ;
}


void clc_timer_unblock(timer_h handle)
{
	timer_s *tp = TP( handle ) ;

	__ostimer_blockall() ;

	if ( tp->t_blocked )
	{
		tp->t_blocked = FALSE ;
		if ( tp->t_act == PENDING )
		{
			tp->t_act = SCHEDULED ;
			(void) __timer_invoke( tp ) ;
		}
	}

	__ostimer_unblockall() ;
}


unsigned clc_timer_expirations(timer_h handle)
{
	return( TP( handle )->t_count ) ;
}



/*
 * Invoke the action of the specified timer
 * All timer interrupts should be blocked when this function is invoked
 * Returns TRUE if
 */
enum timer_state __timer_invoke(register timer_s *tp)
{
	enum timer_state state ;

	/*
	 * The reason for the infinite loop is that the timer may reexpire
	 * while its function is being invoked.
	 */
	for ( ;; )
	{
		/*
		 * This is the INVOKE part
		 */
		if ( tp->t_blocked )
			tp->t_act = PENDING ;
		else
		{
			if ( tp->t_state != DESTROYED && tp->t_act == SCHEDULED )
			{
				void	(*func)()	= tp->t_action.ta_func ;
				void	*arg			= tp->t_action.ta_arg ;
				int	flags 		= tp->t_action.ta_flags ;

				tp->t_act = INVOKED ;
				tp->t_expirations = tp->t_count ;
				tp->t_count = 0 ;
				if ( func != NULL )
				{
					int unblock_all_intrs = ! ( flags & TIMER_BLOCK_ALL ) ;
					int unblock_all_but_same_intr = ! ( flags & TIMER_BLOCK_SAME ) ;
				
					if ( unblock_all_intrs )
						__ostimer_unblockall() ;
					else if ( unblock_all_but_same_intr )
						__ostimer_unblockall_except( tp->t_ostimer ) ;

					(*func)( tp, arg ) ;

					if ( unblock_all_intrs || unblock_all_but_same_intr )
						__ostimer_blockall() ;
				}
				else if ( arg != NULL )
				{
					int *ip = (int *) arg ;

					if ( flags & TIMER_INC_VAR )
						*ip += tp->t_expirations ;
					else
						*ip = 1 ;
				}
			}
		}

		state = tp->t_state ;

		/*
		 * This is the RETURN part
		 */
		if ( tp->t_state == DESTROYED )
			TIMER_FREE( tp ) ;
		else
		{
			if ( tp->t_act == INVOKED )
				tp->t_act = IDLE ;
			else if ( tp->t_act == SCHEDULED )
				continue ;
		}
		break ;
	}
	return( state ) ;
}

